### ISSUE TITLE
Incompatibility with cattrs 23.2, application/json response contents, and `decode_content=True`

### ISSUE THREAD

**Issue by JWCook** [requests-cache/requests-cache#907]
When saving response contents as decoded JSON (with `decode_content=True`), I'm currently seeing errors like this with the latest release of cattrs:
```py
cattrs.errors.StructureHandlerNotFoundError: Unsupported type: typing.Union[typing.List, typing.Dict]. Register a structure hook for it.
Structuring class CachedResponse @ attribute _decoded_content
```

And slightly more confusing:
```py
cattrs.errors.ClassValidationError: While structuring CachedResponse (1 sub-exception)
TypeError: {...} (<class 'dict'>) not part of typing.Union[typing.Dict, typing.List, str, NoneType]
Structuring class CachedResponse @ attribute _decoded_content
```

This might be due to a conflict between the updated [`JsonConverter`](https://github.com/python-attrs/cattrs/blob/main/src/cattrs/preconf/json.py) (using the new ["Union passthrough" strategy](https://catt.rs/en/stable/strategies.html#union-passthrough)) and how it's extended in [`CattrStage.converter`](https://github.com/requests-cache/requests-cache/blob/main/requests_cache/serializers/cattrs.py#L33) to structure `CachedResponse._decoded_content`.

Same results with both `ujson` and stdlib `json`. Like the error message says, it looks like a quick fix for this is to add a new structure hook specifically for `Union[List, Dict]`. There's almost certainly a better way to handle this, though, and I don't quite understand the root cause.

**Comment by JWCook** [comment]
It looks like a quick fix for this is to add a new structure hook for `Union[List, Dict]` (somehow not handled by the existing hook for `Union[Dict, List, str, None]`).

There's almost certainly a better way to handle this, but if I can't figure it out before Monday, I'll make a patch release with that workaround.

**Comment by JWCook** [comment]
Okay, I think I understand what's happening:
* In cattrs: `JsonConverter` uses `configure_union_passthrough()` [here](https://github.com/python-attrs/cattrs/blob/main/src/cattrs/preconf/json.py#L40) to register a Union type containing all JSON primitive types.
    * This is to handle results from a JSON deserializer that produces values with these types. Each value will be structured using its individual type (say, a `float`) instead of treating the Union itself as a special type.
* In requests-cache: `CachedResponse._decoded_content` is annotated with `DecodedContent`, a Union type alias for decoded content from a response with either `application/json` or `text/*`
* In requests-cache: the converter used by `CattrStage` registers a structure hook for `DecodedContent`
* In cattrs: The "union passthrough" strategy kicks in before this hook has a chance to run.
    * It handles JSON primitive types, but not root types (dict and list).
    * The diff between the registered Union and the `DecodedContent` type becomes `Union[List, Dict]`.
    * It handles this as a "spillover type" ([handled here](https://github.com/python-attrs/cattrs/blob/main/src/cattrs/strategies/_unions.py#L169-L196)), and looks for a converter for `Union[List, Dict]`, which doesn't exist, raising an error.

This may be more of a problem with how I'm using cattrs; maybe modifying/extending the pre-configured converters isn't an intended use case? This did work prior to 23.2, though.

**Comment by JWCook** [comment]
So the workaround I mentioned above turns out to be a reasonable solution for this.

I think a better long-term solution would be to implement support for RFC 7159 (#868), and handle the logic for the `decode_content` option *after* the cattrs converter runs, instead of as part of the converter.

**Comment by JWCook** [comment]
Fixed in main, and patched in v1.1.1.
